Repairing the hello_can example.

The hello_can example is a small example that shows the whole cycle of sending and receiving transfers.

It currently doesn't build as a result of recent changes that were needed to be fast, ending up breaking the example itself.

The example will currently not build, with the following error:

```
error[E0277]: the trait bound `ClassicFrame: From<(u32, [u8; 8], usize)>` is not satisfied
  --> examples/hello_can.rs:35:6
   |
35 | impl CanFrame<CLASSIC_MTU> for ClassicFrame {
   |      ^^^^^^^^^^^^^^^^^^^^^ the trait `From<(u32, [u8; 8], usize)>` is not implemented for `ClassicFrame`
   | 
  ::: /home/luca/languages/rust/uavcan/src/lib.rs:9:39
   |
9  | pub trait CanFrame<const MTU: usize>: From<(u32, [u8; MTU], usize)> + core::fmt::Debug {
   |                                       ----------------------------- required by this bound in `CanFrame`
   |
   = help: the following implementations were found:
             <ClassicFrame as From<(u32, [u8; 8])>>
             <ClassicFrame as From<CANFrame>>

error[E0277]: the trait bound `ClassicFrame: From<(u32, [u8; 8], usize)>` is not satisfied
  --> examples/hello_can.rs:47:6
   |
47 | impl CanWriter<ClassicFrame, CLASSIC_MTU> for CanTx {
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<(u32, [u8; 8], usize)>` is not implemented for `ClassicFrame`
   | 
  ::: /home/luca/languages/rust/uavcan/src/tx/stream_transmitter.rs:6:28
   |
6  | pub trait CanWriter<Frame: CanFrame<MTU>, const MTU: usize> {
   |                            ------------- required by this bound in `CanWriter`
   |
   = help: the following implementations were found:
             <ClassicFrame as From<(u32, [u8; 8])>>
             <ClassicFrame as From<CANFrame>>

```

This is the result of recently changing the required CanFrame trait to interface.
In particular, the interface initially only required a conversion from `(u32, [u8; 8])`, as it was expected that each frame always contained the whole MTU amount of data.

This was later changed to take into account receiving frames that are partially filled.
For example, supposing that the payload that is being transmitted fill exactly `N` frames of data, the last frame, that will contain the CRC, should not contain the whole amount of data, but only the two bytes of the CRC followed by the closing tail byte.

To support this behavior, the extraction of the tail byte of a received frame required knowledge about the DMC of the received frame, to identify where the tail byte is located and what part of the frame is the actual payload.
The interface was thus changed to depend on a further `usize` parameter representing the length of the data in the frame.

The implementation of the `CanFrame` trait should be updated to the new one to allow the example to be compiled again.

--- a/examples/hello_can.rs
+++ b/examples/hello_can.rs
@@ -13,11 +13,12 @@ use uavcan::{CanFrame, CLASSIC_MTU};
 struct ClassicFrame {
     data: [u8; 8],
     id: u32,
+    len: usize,
 }
 
-impl From<(u32, [u8; 8])> for ClassicFrame {
-    fn from((id, data): (u32, [u8; 8])) -> Self {
-        Self { data, id }
+impl From<(u32, [u8; 8], usize)> for ClassicFrame {
+    fn from((id, data, len): (u32, [u8; 8], usize)) -> Self {
+        Self { data, id, len }
     }
 }
 
@@ -28,7 +29,7 @@ impl From<CANFrame> for ClassicFrame {
         let mut data = [0u8; 8];
         data[..original_data.len()].copy_from_slice(original_data);
 
-        Self::from((id, data))
+        Self::from((id, data, original_data.len()))
     }
 }
 
@@ -37,8 +38,8 @@ impl CanFrame<CLASSIC_MTU> for ClassicFrame {
         self.id
     }
 
-    fn payload(&self) -> &[u8; CLASSIC_MTU] {
-        &self.data
+    fn payload(&self) -> (&[u8; CLASSIC_MTU], usize) {
+        (&self.data, self.len)
     }
 }
 
@@ -48,8 +49,9 @@ impl CanWriter<ClassicFrame, CLASSIC_MTU> for CanTx {
     type Error = std::io::Error;
 
     fn write_frame(&mut self, frame: ClassicFrame) -> Result<(), Self::Error> {
+        let (payload, payload_len) = frame.payload();
         self.0
-            .write_frame(&CANFrame::new(frame.id(), frame.payload(), false, false).unwrap())
+            .write_frame(&CANFrame::new(frame.id(), &payload[..payload_len], false, false).unwrap())
     }
 }
 

